telemetry/csharp/ToolkitTelemetryGenerator/DefinitionsBuilder.cs

```diff
@@ -0,0 +1,757 @@
+        private CodeNamespace _blankNamespace; // Used for top level Using statements
+        private CodeNamespace _generatedNamespace; // Where generated code goes
+
+        private CodeTypeDeclaration _telemetryEventsClass;
+
+        public DefinitionsBuilder AddMetricsTypes(IList<MetricType> types)
+        {
+            _types.AddRange(types);
+
+            return this;
+        }
+
+        public DefinitionsBuilder AddMetrics(IList<Metric> metrics)
+        {
+            _metrics.AddRange(metrics);
+
+            return this;
+        }
+
+        public DefinitionsBuilder WithNamespace(string generatedNamespace)
+        {
+            _namespace = generatedNamespace;
+
+            return this;
+        }
+
+        public string Build()
+        {
+            if (string.IsNullOrWhiteSpace(_namespace))
+            {
+                throw new MissingFieldException("Namespace not provided");
+            }
+
+            _generatedCode = new CodeCompileUnit();
+            _blankNamespace = new CodeNamespace();
+            _generatedNamespace = new CodeNamespace(_namespace);
+
+            _generatedCode.Namespaces.Add(_blankNamespace);
+            _generatedCode.Namespaces.Add(_generatedNamespace);
+
+            // Add a top level comment
+            _blankNamespace.Comments.Add(new CodeCommentStatement("--------------------------------------------------------------------------------", true));
+            _blankNamespace.Comments.Add(new CodeCommentStatement("This file is generated from https://github.com/aws/aws-toolkit-common/tree/master/telemetry", true));
+            _blankNamespace.Comments.Add(new CodeCommentStatement("--------------------------------------------------------------------------------", true));
+
+            // Set up top level using statements
+            _blankNamespace.Imports.Add(new CodeNamespaceImport("System"));
+            _blankNamespace.Imports.Add(new CodeNamespaceImport("System.Collections.Generic"));
+            _blankNamespace.Imports.Add(new CodeNamespaceImport("log4net"));
+
+            // public sealed class ToolkitTelemetryEvent (contains generated code the toolkit uses to record metrics)
+            _telemetryEventsClass = new CodeTypeDeclaration()
+            {
+                Name = "ToolkitTelemetryEvent",
+                IsClass = true,
+                TypeAttributes = TypeAttributes.Public
+            };
+            _telemetryEventsClass.Comments.Add(new CodeCommentStatement("Contains methods to record telemetry events", true));
+            _generatedNamespace.Types.Add(_telemetryEventsClass);
+
+            GenerateFixedCode();
+            ProcessMetricTypes();
+            ProcessMetrics();
+
+            // Output generated code to a string
+            CodeDomProvider provider = CodeDomProvider.CreateProvider("CSharp");
+            CodeGeneratorOptions options = new CodeGeneratorOptions
+            {
+                BracingStyle = "C", 
```

Comment:
wow they really got this backwards, I was assuming this meant brackets on the same line. I had to look it up